airflow/api_fastapi/execution_api/routes/task_instance.py

```diff
@@ -0,0 +1,132 @@
+    """
+    # We only use UUID above for validation purposes
+    ti_id_str = str(task_instance_id)
+
+    old = select(TI.state).where(TI.id == ti_id_str).with_for_update()
+    try:
+        (previous_state,) = session.execute(old).one()
+    except NoResultFound:
+        log.error("Task Instance %s not found", ti_id_str)
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail={
+                "reason": "not_found",
+                "message": "Task Instance not found",
+            },
+        )
+
+    # We exclude_unset to avoid updating fields that are not set in the payload
+    data = ti_patch_payload.model_dump(exclude_unset=True)
+
+    query = update(TI).where(TI.id == ti_id_str).values(data)
+
+    if isinstance(ti_patch_payload, schemas.TIEnterRunningPayload):
+        if previous_state != State.QUEUED:
+            log.warning(
+                "Can not start Task Instance ('%s') in invalid state: %s",
+                ti_id_str,
+                previous_state,
+            )
+
+            # TODO: Pass a RFC 9457 compliant error message in "detail" field
+            # https://datatracker.ietf.org/doc/html/rfc9457
+            # to provide more information about the error
+            # FastAPI will automatically convert this to a JSON response
+            # This might be added in FastAPI in https://github.com/fastapi/fastapi/issues/10370
+            raise HTTPException(
+                status_code=status.HTTP_409_CONFLICT,
+                detail={
+                    "reason": "invalid_state",
+                    "message": "TI was not in a state where it could be marked as running",
+                    "previous_state": previous_state,
+                },
+            )
+        log.info("Task with %s state started on %s ", previous_state, ti_patch_payload.hostname)
+        # Ensure there is no end date set.
+        query = query.values(
+            end_date=None,
+            hostname=ti_patch_payload.hostname,
+            unixname=ti_patch_payload.unixname,
+            pid=ti_patch_payload.pid,
+            state=State.RUNNING,
+        )
+    elif isinstance(ti_patch_payload, schemas.TITerminalStatePayload):
+        query = TI.duration_expression_update(ti_patch_payload.end_date, query, session.bind)
+
+    # TODO: Replace this with FastAPI's Custom Exception handling:
+    # https://fastapi.tiangolo.com/tutorial/handling-errors/#install-custom-exception-handlers
+    try:
+        result = session.execute(query)
+        log.info("TI %s state updated: %s row(s) affected", ti_id_str, result.rowcount)
+        session.commit()
+    except SQLAlchemyError as e:
+        session.rollback()
+        log.error("Error updating Task Instance state: %s", e)
+        raise HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Database error occurred"
```

Comment:
I think the rollback on `SQLAlchemyError` is already done by the `get_session` context manager. (This is inherent to all endpoints, all of them should rollback the transaction if something went wrong in the endpoint)

Reply:
It should rollback on any error, right, so this would work?

Do we need the `session.commit()` too, or is that handled for us as well?
```suggestion
        result = session.execute(query)
        log.info("TI %s state updated: %s row(s) affected", ti_id_str, result.rowcount)
        session.commit()
    except SQLAlchemyError as e:
        log.error("Error updating Task Instance state: %s", e)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Database error occurred"
```

Reply:
Removed session.commit & session.rollback but kept the error handling to raise HTTPException for now (for this PR)